ui/src/onboarding/containers/LoginPageContents.tsx

```diff
@@ -0,0 +1,383 @@
+
+    return {isValid, errors}
+  }
+
+  private handleSubmit = (e: FormEvent) => {
+    const {isValid, errors} = this.validFieldValues
+    const {
+      email,
+      password,
+      firstName: given_name,
+      lastName: family_name,
+      activeTab,
+    } = this.state
+
+    e.preventDefault()
+
+    if (!isValid) {
+      this.setState(errors)
+      return
+    }
+
+    this.setState({buttonStatus: ComponentStatus.Loading})
+
+    if (activeTab === 'login') {
+      auth0.login(
+        {
+          realm: Auth0Connection.Authentication,
+          email,
+          password,
+        },
+        err => {
+          if (err) {
+            this.setState({buttonStatus: ComponentStatus.Default})
+            return this.displayErrorMessage(errors, err)
+          }
+        }
+      )
+      return
+    }
+
+    auth0.signup(
+      {
+        connection: Auth0Connection.Authentication,
+        email,
+        password,
+        family_name,
+        given_name,
+      },
+      err => {
+        if (err) {
+          this.displayErrorMessage(errors, err)
+          this.setState({buttonStatus: ComponentStatus.Default})
+          return
+        }
+        // log the user into Quartz
+        auth0.login(
+          {
+            realm: Auth0Connection.Authentication,
+            email,
+            password,
+          },
+          error => {
+            if (error) {
+              this.setState({buttonStatus: ComponentStatus.Default})
+              return this.displayErrorMessage(errors, error)
+            }
+          }
+        )
+      }
+    )
+  }
+
+  private displayErrorMessage = (errors, auth0Err) => {
+    if (auth0Err.code.includes('error in email')) {
+      this.setState({
+        ...errors,
+        emailError: 'Please enter a valid email address',
+      })
+    } else if (auth0Err.code === 'user_exists') {
+      const emailError = `An account with that email address already exists. Try logging in instead.`
+      this.setState({...errors, emailError})
+    } else {
+      const emailError = `We have been notified of an issue while creating your account. If this issue persists, please contact support@influxdata.com`
+      this.setState({...errors, emailError})
+      throw new Error(auth0Err.description)
+    }
+  }
+
+  private handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
+    this.setState({[e.target.name]: e.target.value})
+  }
+
+  private handleTabChange = (value: string) => {
+    this.setState({activeTab: value})
+  }
+
+  private handleSocialClick = (connection: Auth0Connection) => {
+    auth0.authorize({
+      connection,
+    })
+  }
+
+  private handleForgotPasswordClick = () => {
+    const {email} = this.state
+    if (!email) {
+      this.setState({emailError: 'Please enter a valid email address'})
+      return
+    }
+    auth0.changePassword(
+      {
+        email,
+        connection: Auth0Connection.Authentication,
+      },
+      (err, successMessage) => {
```

Comment:
this callback style is kind of gnarly, and gets kind of deep in some code above. did you look into wrapping this library in a shallow wrapper to make use of promises, which is pretty much how we do async code throughout the codebase?

Reply:
Totally agree, this is based on the docs, but I can see if it's possible to refactor these into async awaits

Reply:
This is more a result of the design of the API where the second parameter that's accepted is the callback. I tried refactoring this to promisify the result but didn't have the same outcome